### ISSUE TITLE
Question: How to adjust speed of clip?

### ISSUE THREAD

**Issue by jeffski** [OpenShot/libopenshot#176]
This isn't an issue but hoping someone could give me a code example or create an example (maybe I can do that if I work this out).

I am trying to speed up and slow down and reverse a clip. I believe this should be possible using the time property of a clip: 

`Keyframe time; ///< Curve representing the frames over time to play (used for speed and direction of video)`

But I can't work out how to apply this to a clip. Can some one please provide me with some code examples or suggestions.

**Comment by peanutbutterandcrackers** [comment]
I wonder if this is related in any ways to this: https://github.com/OpenShot/openshot-qt/issues/1198 (for which an explanation has been made here: https://github.com/OpenShot/openshot-qt/issues/1032#issuecomment-408429421).

Pinging @N3WWN as I think he can help you with whatever the issue here is.

**Comment by N3WWN** [comment]
Depending on how many of these "effects" you want to apply to a single clip will determine how many steps it will take to perform each one.

(Note:  My machine which I use for video processing and Openshot development is currently down awaiting a new power supply to arrive, so this is all from memory and may not be entirely correct, but will hopefully get you started down the right path.)

To speed up an entire clip using the preset values, you can right-click on the clip within the timeline, select `Time` -> `Fast` -> `Forward` -> `2x` (or whatever speed you want).  The same thing works for slowing down the entire clip or reversing the entire clip at these preset rates.

To change the rate to a non-preset value, you need to determine how many frames the clip contains at the current project rate.  

If I opened a clip that is `1min17sec14frames` long into a 30fps project, so that clip is 2324 frames in duration (`(1min*60+17sec)*30+14`).  If I want that clip to play 10% faster, it needs to have a duration that is 232.4 frames shorter than 2324 frames, or 2092 frames.  2092 frames / 30 fps = 69 sec (`1min9sec`), so I move my playhead to 1min9sec1frame, make sure the clip is selected, and enter `2324` in the `Time` property.

There will be a green tick at the bottom of the clip where this keyframe was set.  The clip will extend beyond this keyframe, but will just be the last frame frozen until the end of the original duration of the clip.  To get rid of this, you can right-click with the playhead on the keyframe, `Slice` -> `Keep Left Side`.

Manually reversing the clip with a non-preset speed is the same, but you set two keyframes:  the one at the start of the clip contains the last frame number and the one at the end of the clip contains the first frame number.  In the above example, if I wanted to play this clip 10% faster in reverse, the keyframe at the start of the clip would have a `Time` value of `2324` and the keyframe at the end of the clip would be `1`.

Technically, two keyframes are created during the non-preset forward rate change because the first frame is automagically set as a keyframe with a value of `1`.

You can set other keyframes on the timeline for a single clip, too.

So, with my 10% faster forward clip, I could set keyframes at the first and last frames of the clip with a `Time` value of `2324`, and a third at, say `0min34sec30frames`, with a `Time` value of `1`.  This would play the entire clip backwards at double the 10% faster speed from frame `0min0sec1frame` to `0min34sec30frames`, reverse direction and play the entire clip forwards at double the 10% faster speed from frame `0min34sec30frames` until `1min9sec1frames`.

You may need to apply the values to the keyframes multiple times since OpenShot appears to try to keep some values "sane" (or what it thinks are sane) and may overwrite your initial value(s) when you add the next keyframe(s).

If the change in frame rate mid-clip doesn't feel right, you can right-click on the little chart to the right of the value in the `Time` property and change to a Bezier easing (see also [Easing Functions](https://easings.net)).  

In my last example, the output would jump from double the 10% speed reverse to double the 10% speed forwards, but it would feel smoother (or like you were reversing a film projector) if you had two keyframes with a `Time` value of `1` (`0min34sec30frames` and `0min35sec1frames`) changed to a Bezier `Ease In` for the first one and `Ease Out` for the second one.  I know it seems like you should be able to set `Ease In/Out` for the single keyframe, but from what I recall, that doesn't work with rate changes.  

Playing with the `Time` properties to get a desired effect can be difficult at first, but the results can be really great!

**Comment by Noino** [comment]
TL;DR guide to the comprehensive guide by @N3WWN
https://github.com/OpenShot/openshot-qt/issues/327#issuecomment-463998686

**Comment by poszy6** [comment]
For me this isnt working at all. When I seleck the time -> fast -> forward -> 2X (or any of the values) the clip shrinks, but when I try to play it, it froze and lag.  I think when you use 2X(or other) OpenShot calculate the new time value for you ( its just a hunch ).

**Comment by ferdnyc** [comment]
@poszy6 That'll likely depend on the input video. Speed changes work flawlessly on some videos with no detectable impact on performance. On other files they can cause huge problems, or fail completely. That's something that's probably out of OpenShot's hands, for the most part. The best way to solve it is usually to convert the video into a cleaner format with better support for things like seeking and frame rate changes.

Videos created in real time by a live-capture process (whether it's from a camera, a stream recorder, or etc.) tend to be poorly-indexed and encoded sub-optimally, and they'll often lead to these problems.

Creating video files using an "offline" process, where the entire file is provided from start to finish, means that the software is able to make intelligent decisions about how to optimally format the resulting file. Files created that way will typically just work far better. Especially if a container format like MKV (Matroska) is used, since it was designed to be as accommodating as possible to subsequent processing and adjustment of the contents.

(Note that a video doesn't _necessarily_ have to be transcoded to be more compatible with the editing process. Often, simply remuxing the original file's encoded parts into a new MKV file will be enough. Remuxing is enough to create the metadata and indexing that the original may lack, which is frequently the biggest source of performance issues during playback and editing.)

As I said, though, there's really not that much OpenShot could do about that, other than just automatically converting the video _for_ the user right from the start. (Which is a possibility, one that's been discussed from time to time.) Compressed videos encoded on-the-fly with poor indexing are never **not** going to cause problems when trying to use them as input files.
